src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -2600,5 +2623,273 @@ private void Remove(int startIndex, int count, out StringBuilder chunk, out int
+            public static InterpolatedAppendBuilder Create(int literalLength, int formattedCount, StringBuilder stringBuilder, IFormatProvider? provider) =>
+                new InterpolatedAppendBuilder(stringBuilder, provider);
+
+            /// <summary>Writes the specified string to the builder.</summary>
+            /// <param name="value">The string to write.</param>
+            public void AppendLiteral(string value) => _stringBuilder.Append(value);
+
+            #region AppendFormatted
+            // Design note:
+            // This provides the same set of overloads and semantics as InterpolatedStringBuilder.
+
+            #region AppendFormatted T
+            /// <summary>Writes the specified value to the builder.</summary>
+            /// <param name="value">The value to write.</param>
+            public void AppendFormatted<T>(T value) => AppendFormatted(value, format: null);
+
+            /// <summary>Writes the specified value to the builder.</summary>
+            /// <param name="value">The value to write.</param>
+            /// <param name="format">The format string.</param>
+            public void AppendFormatted<T>(T value, string? format)
+            {
+                if (_hasCustomFormatter)
+                {
+                    // If there's a custom formatter, always use it.
+                    AppendCustomFormatter(value, format);
+                }
+                else if (value is IFormattable)
+                {
+                    // Check first for IFormattable, even though we'll prefer to use ISpanFormattable, as the latter
+                    // requires the former.  For value types, it won't matter as the type checks devolve into
+                    // JIT-time constants.  For reference types, they're more likely to implement IFormattable
+                    // than they are to implement ISpanFormattable: if they don't implement either, we save an
+                    // interface check over first checking for ISpanFormattable and then for IFormattable, and
+                    // if it only implements IFormattable, we come out even: only if it implements both do we
+                    // end up paying for an extra interface check.
+
+                    if (value is ISpanFormattable)
+                    {
+                        Span<char> destination = _stringBuilder.RemainingCurrentChunk;
+                        if (((ISpanFormattable)value).TryFormat(destination, out int charsWritten, format, _provider)) // constrained call avoiding boxing for value types
+                        {
+                            if ((uint)charsWritten > (uint)destination.Length)
+                            {
+                                // Protect against faulty ISpanFormattable implementations returning invalid charsWritten values.
+                                FormatError();
```

Comment:
I don't think I saw a test that hits this.  Either from a bad implementation reporting a negative number, or too big of one.

Reply:
Good point.  I'll add that.